### PR TITLE
Inhibit direct conversion from future<future<T>> --> future<void>

### LINKED ISSUE #2667: Ambiguity of nested hpx::future<void>'s.
When future < T >  is created using a function returning a future < T >  instead of a T, the outer future is not made ready until the inner future is complete.

However, when a futue< void > is created using a function returning a future< void > instead of void, the outer future is made ready as soon as the inner future is returned - which could be (and often is) before the inner future is made ready. 

Example:
```
hpx::future<void> fut = hpx::async( []() {
     do_work();
     return hpx::async( []() {
        do_more_work();     
     };
});
```
Given the behavior of nested non-void futures, one would expect fut would not be made ready until after do_more_work() is executed. Apparently (from bugs I have fixed in Octo-tiger) this is not the case. Instead fut is made ready when the inner future is created and returned, which is (usually) before do_more_work() is complete.

I have fixed two bugs in Octo-tiger that were of this type, and each time the bugs were not readily reproducible and took many days and SUs to narrow down the problem and confirm I had fixed it. 

Is this the intended behavior of nested void futures or is this a bug?

### PR DESCRIPTION
**Pull request by hkaiser**
- adding corresponding tests

- this fixes #2667

### CODE REVIEW

**Comment by hkaiser** [comment]
@dmarce1 Please check this out and let us know if this fixes the issue you're seeing.

File: tests/regressions/lcos/fail_future_2667.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        hpx::async(
+            []() -> hpx::future<int> {
+                return hpx::async(
+                    []() -> int {
+                        return 42;
+                    });
+            });
+
+    hpx::future<void> fut2 = std::move(fut);
+    fut2.get();
```
Comment (sithhell) [inline tests/regressions/lcos/fail_future_2667.cpp]: Shouldn't we test here that the inner async has been executed after fut2.get() succeeded?

File: tests/regressions/lcos/future_2667.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                        do_more_work();
+                    });
+            });
+
+    hpx::util::high_resolution_timer t;
+
+    hpx::future<void> fut2 = std::move(fut);
+    fut2.get();
+
+    HPX_TEST(t.elapsed() > 1.0);
```
Comment (sithhell) [inline tests/regressions/lcos/future_2667.cpp]: Is checking the elapsed time the right thing to do or shouldn't we rather check if some flag or similar has been set?

Comment (hkaiser) [inline tests/regressions/lcos/future_2667.cpp]: Yes, checking a flag can be done additionally. The time check is supposed to make sure that the collapsed future becomes ready only after the inner future.

File: tests/regressions/lcos/fail_future_2667.cpp (hunk 1 above)
Comment (hkaiser) [inline tests/regressions/lcos/fail_future_2667.cpp]: This is a test making sure that things don't compile. No testing of flags would make a difference.

**Comment by hkaiser** [comment]
@sithhell your comments have been addressed.

File: tests/regressions/lcos/fail_future_2667.cpp (hunk 1 above)
Comment (sithhell) [inline tests/regressions/lcos/fail_future_2667.cpp]: ahh, missed that this is just a compile only test ... giving essentially the same test below, is a separate, compile only test really necessary? The test will fail if it does not compile.

Comment (hkaiser) [inline tests/regressions/lcos/fail_future_2667.cpp]: It's not the same test. The compile-fail test makes sure the conversion `future<future<T>> -> future<void>` does not compile (`T` != `void`), while the other test makes sure the conversion `future<future<void>> -> future<void>` is doing what it is supposed to do, namely unwrapping the nested future.

**Comment by hkaiser** [comment]
@dmarce1 Thanks. We know for sure that this PR addresses a problem, it might not be the problem you were seeing. With this PR, did you see compilation errors with octotiger? Did you have to change anything?

File: tests/regressions/lcos/future_2667.cpp (hunk 2 above)
Comment (hkaiser) [inline tests/regressions/lcos/future_2667.cpp]: @sithhell I now have added the additional check as you suggested.

**Comment by dmarce1** [comment]
@hkaiser  No compilation errors, didn't have to change anything. Everything appears to be working.

**Comment by hkaiser** [comment]
> No compilation errors, didn't have to change anything. Everything appears to be working.

@dmarce1: ok, then you're most likely not affected by the problem this PR solves.

**Pull request merged**